File: Backend/PicturesAPI/app/models/history.py
"""
Project: PicturesAPI
Author: Diego C. <20026893@tafe.wa.edu.au>
Created at: 28/09/2020 8:57 pm
File: history.py
"""
from sqlalchemy import func, Column, BIGINT, DateTime, ForeignKey
from sqlalchemy.orm import relationship

from app.database.db import Base
from .face_shape import FaceShape
from .hair_colour import HairColour
from .hair_style import HairStyle
from .picture import Picture
from .user import User


class History(Base):
    __tablename__ = "history"
    id = Column(BIGINT, primary_key=True, index=True)
    picture_id = Column(BIGINT, ForeignKey("pictures.id"), index=True)
    original_picture_id = Column(BIGINT, ForeignKey("pictures.id"), index=True)
    previous_picture_id = Column(BIGINT, ForeignKey("pictures.id"), index=True)
    hair_colour_id = Column(BIGINT, ForeignKey("colours.id"), index=True)
    hair_style_id = Column(BIGINT, ForeignKey("hair_styles.id"), index=True)
    face_shape_id = Column(BIGINT, ForeignKey("face_shapes.id"), index=True)
    user_id = Column(BIGINT, ForeignKey("users.id"), index=True)
    date_created = Column(DateTime(timezone=True), server_default=func.now())
    date_updated = Column(DateTime(timezone=True), onupdate=func.now())

    picture = relationship("Picture", back_populates="history", foreign_keys=[picture_id])
    original_picture = relationship("Picture", back_populates="original_history", foreign_keys=[original_picture_id])
    previous_picture = relationship("Picture", back_populates="previous_history", foreign_keys=[previous_picture_id])
    hair_colour = relationship("HairColour", back_populates="history")
    hair_style = relationship("HairStyle", back_populates="history")
    face_shape = relationship("FaceShape", back_populates="history")
    user = relationship("User", back_populates="history")

    def __repr__(self):
        return 'History entry:\n' + \
               f'id: {self.id}\n' + \
               f'picture_id: {self.picture_id}\n' + \
               f'original_picture_id: {self.original_picture_id}\n' + \
               f'previous_picture_id: {self.previous_picture_id}\n' + \
               f'hair_colour_id: {self.hair_colour_id}\n' + \
               f'hair_style_id: {self.hair_style_id}\n' + \
               f'face_shape_id: {self.face_shape_id}\n' + \
               f'user_id: {self.user_id}\n' + \
               f'date_created: {self.date_created}\n' + \
               f'date_updated: {self.date_updated}\n'


Picture.history = relationship(
    "History", back_populates="picture", foreign_keys=[History.picture_id], cascade="all, delete-orphan, save-update"
)

Picture.original_history = relationship(
    "History", back_populates="original_picture", foreign_keys=[History.original_picture_id],
    cascade="all, delete-orphan, save-update"
)

Picture.previous_history = relationship(
    "History", back_populates="previous_picture", foreign_keys=[History.previous_picture_id],
    cascade="all, delete-orphan, save-update"
)

User.history = relationship(
    "History", order_by=History.date_updated, back_populates="user", cascade="all, delete-orphan"
)

HairColour.history = relationship(
    "History", back_populates="hair_colour", cascade="all, delete-orphan, save-update"
)

HairStyle.history = relationship(
    "History", back_populates="hair_style", cascade="all, delete-orphan, save-update"
)

FaceShape.history = relationship(
    "History", back_populates="face_shape", cascade="all, delete-orphan, save-update"
)

File: Backend/UsersAPI/Models/History.cs
using System;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace UsersAPI.Models
{
    [Table("history")]
    public partial class History
    {
        [Key]
        [Column("id", TypeName = "bigint(20) unsigned")]
        public ulong Id { get; set; }
        [Column("picture_id", TypeName = "bigint(20) unsigned")]
        public ulong PictureId { get; set; }
        [Column("original_picture_id", TypeName = "bigint(20) unsigned")]
        public ulong? OriginalPictureId { get; set; }
        [Column("previous_picture_id", TypeName = "bigint(20) unsigned")]
        public ulong? PreviousPictureId { get; set; }
        [Column("hair_colour_id", TypeName = "bigint(20) unsigned")]
        public ulong? HairColourId { get; set; }
        [Column("hair_style_id", TypeName = "bigint(20) unsigned")]
        public ulong? HairStyleId { get; set; }
        [Column("face_shape_id", TypeName = "bigint(20) unsigned")]
        public ulong? FaceShapeId { get; set; }
        [Column("user_id", TypeName = "bigint(20) unsigned")]
        public ulong UserId { get; set; }
        [Column("date_created", TypeName = "datetime")]
        public DateTime? DateCreated { get; set; }
        [Column("date_updated", TypeName = "datetime")]
        public DateTime? DateUpdated { get; set; }

        [ForeignKey(nameof(FaceShapeId))]
        [InverseProperty(nameof(FaceShapes.History))]
        public virtual FaceShapes FaceShape { get; set; }
        [ForeignKey(nameof(HairColourId))]
        [InverseProperty(nameof(Colours.History))]
        public virtual Colours HairColour { get; set; }
        [ForeignKey(nameof(HairStyleId))]
        [InverseProperty(nameof(HairStyles.History))]
        public virtual HairStyles HairStyle { get; set; }
        [ForeignKey(nameof(OriginalPictureId))]
        [InverseProperty(nameof(Pictures.HistoryOriginalPicture))]
        public virtual Pictures OriginalPicture { get; set; }
        [ForeignKey(nameof(PictureId))]
        [InverseProperty(nameof(Pictures.HistoryPicture))]
        public virtual Pictures Picture { get; set; }
        [ForeignKey(nameof(PreviousPictureId))]
        [InverseProperty(nameof(Pictures.HistoryPreviousPicture))]
        public virtual Pictures PreviousPicture { get; set; }
        [ForeignKey(nameof(UserId))]
        [InverseProperty(nameof(Users.History))]
        public virtual Users User { get; set; }
    }
}
